### PR TITLE
Create parent by default when copy object in Rest API

### PR DESCRIPTION
**Pull request by Jackson-Wang-7**
### What changes are proposed in this pull request?

setRecursive to true in Copy Object request.

### Why are the changes needed?

if it’s a copy request, we may need to create the parent by default in the Object Storage standard.

### CODE REVIEW

File: tests/src/test/java/alluxio/client/rest/S3ClientRestApiTest.java [hunk 1]
```diff
@@ -1535,6 +1535,46 @@ public void testTaggingNoLimit() throws Exception {
...
+
+    String object = CommonUtils.randomAlphaNumString(DATA_SIZE);
+    final String fullObjectKey = bucketName + AlluxioURI.SEPARATOR + objectKey;
+    String copiedObjectKey = bucketName + targetObject;
+    AlluxioURI copiedObjectURI = new AlluxioURI(AlluxioURI.SEPARATOR + copiedObjectKey);
+
+    createBucketRestCall(bucketName);
+    createObject(fullObjectKey, object.getBytes(), null, null);
+
+    // metadata directive = REPLACE, tagging directi
```
Comment (ZhuTopher) [inline tests/src/test/java/alluxio/client/rest/S3ClientRestApiTest.java:1552]: nit: Typo? Neither directive affect this tests anyway though.

File: tests/src/test/java/alluxio/client/rest/S3ClientRestApiTest.java [hunk 2]
```diff
@@ -1535,6 +1535,46 @@ public void testTaggingNoLimit() throws Exception {
     testTagging(objectKey, ImmutableMap.copyOf(tagMap));
   }
 
+  @Test
+  public void testCopyObject() throws Exception {
+    final String bucketName = "bucket";
+    final String objectKey = "object";
+    final String targetObject = "/tmp/target";
```
Comment (ZhuTopher) [inline tests/src/test/java/alluxio/client/rest/S3ClientRestApiTest.java:1542]: I would like to add a description to this test (comments) to explain what we're checking is that you can copy an object from one folder to a different folder, and those sub-directories will get created.

**Review by ZhuTopher — APPROVED** [review]
For the sake of expediting the PR, I don't think the comments need to be addressed

**Comment by ZhuTopher** [comment]
# Testing

1. Create a serverless EMR cluster via [AWS console/cli](https://docs.aws.amazon.com/emr/latest/EMR-Serverless-UserGuide/getting-started.html#gs-console)
- I just used the AWS console, created a low-spec 1-master 1-worker EMR w/ Spark pre-installed.

2. Create a standalone Alluxio cluster in EC2 into the same VPC as the EMR cluster

3. On the Alluxio master, create the following paths/files:
```
$ sudo su alluxio && cd /opt/alluxio
echo "John Smith,jsmith@email.com,555-1212" > customers.csv
echo "Janis Joplin,jjoplin@email.com,555-1212" >> customers.csv

$ ./bin/alluxio fs mkdir /testbucket

$ aws configure
AWS Access Key ID [None]: alluxio
AWS Secret Access Key [None]: dummy
Default region name [None]: 
Default output format [None]: 
$ aws s3 --endpoint "http://localhost:39999/api/v1/s3" cp ./customers.csv s3://testbucket/test.csv
upload: ./customers.csv to s3://testbucket/test.csv
```
- You have to use the S3 API to upload objects or else they will fail for having no ETags (this behaviour is being patched out, but until then you must do this)

4. Add the EMR cluster (master AND worker) to the Alluxio SecurityGroup ingress, and vice-versa.
- IMPORTANT: you will need to remove these ingress rules in order for Terraform to teardown the cluster afterwards

5. Run the following commands in a spark shell on the EMR master:

```
spark-shell \
     --driver-library-path /usr/lib/hadoop-lzo/lib/native \
     --jars "/usr/lib/hadoop-lzo/lib/hadoop-lzo.jar,/usr/lib/hadoop/hadoop-aws.jar,$(echo /usr/share/aws/aws-java-sdk/*.jar | tr ' ' ',')"
```

```
val hadoop_conf = spark.sparkContext.hadoopConfiguration
hadoop_conf.set("fs.s3a.impl", "org.apache.hadoop.fs.s3a.S3AFileSystem")
hadoop_conf.set("fs.s3a.access.key", "alluxio")
hadoop_conf.set("fs.s3a.secret.key", "dummy value")
hadoop_conf.set("fs.s3a.path.style.access", "true")
hadoop_conf.set("fs.s3a.endpoint", "http://{ALLUXIO_MASTER_IP}:39999/api/v1/s3")

val df = spark.read.csv("s3a://testbucket/customer.csv")
df.show()
df.rdd.saveAsTextFile("s3a://testbucket/copy.csv")
```

Expected output:
```
scala> val df = spark.read.csv("s3a://testbucket/customer.csv")
df: org.apache.spark.sql.DataFrame = [_c0: string, _c1: string ... 1 more field]

scala> df.show()
+------------+-----------------+--------+
|         _c0|              _c1|     _c2|
+------------+-----------------+--------+
|  John Smith| jsmith@email.com|555-1212|
|Janis Joplin|jjoplin@email.com|555-1212|
+------------+-----------------+--------+
                                                                                
scala> val df2 = spark.read.csv("s3a://testbucket/out.csv")
df2: org.apache.spark.sql.DataFrame = [_c0: string, _c1: string ... 1 more field]

scala> df2.show()
+-------------+-----------------+---------+
|          _c0|              _c1|      _c2|
+-------------+-----------------+---------+
|  [John Smith| jsmith@email.com|555-1212]|
|[Janis Joplin|jjoplin@email.com|555-1212]|
+-------------+-----------------+---------+
```

File: tests/src/test/java/alluxio/client/rest/S3ClientRestApiTest.java (hunk 1 above)
Comment (Jackson-Wang-7) [reply tests/src/test/java/alluxio/client/rest/S3ClientRestApiTest.java]: fixed

File: tests/src/test/java/alluxio/client/rest/S3ClientRestApiTest.java (hunk 2 above)
Comment (Jackson-Wang-7) [reply tests/src/test/java/alluxio/client/rest/S3ClientRestApiTest.java]: added

**Comment by yyongycy** [comment]
alluxio-bot, merge this please

**Comment by yyongycy** [comment]
alluxio-bot, merge this please

**Comment by ZhuTopher** [comment]
alluxio-bot, cherry-pick this to branch-2.8 please

**Comment by ZhuTopher** [comment]
alluxio-bot, cherry-pick this to branch-2.8 please

**Pull request merged**